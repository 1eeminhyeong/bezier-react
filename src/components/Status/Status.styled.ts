/* Internal dependencies */
import { styled, absoluteCenter, SemanticNames } from 'Foundation'
import { Icon } from 'Components/Icon'
import { StatusSize } from './Status.types'

function getStatusCircleBorderSize(size: StatusSize) {
  if (size >= StatusSize.L) { return 3 }
  return 2
}

interface StatusCircleProps {
  color: SemanticNames
  size: StatusSize
}

export const StatusCircle = styled.div<StatusCircleProps>`
  ${({ foundation, size }) => foundation?.border?.getBorder(getStatusCircleBorderSize(size), foundation?.theme?.['bdr-white'])};

  position: relative;
  box-sizing: content-box;
  width: ${({ size }) => size}px;
  height: ${({ size }) => size}px;
  background-color: ${({ foundation }) => foundation?.theme?.['bg-white-normal']};
  border-radius: 50%;

  &::after {
    position: absolute;
    top: 0;
    left: 0;
    display: block;
    width: ${({ size }) => size}px;
    height: ${({ size }) => size}px;
    content: '';
    background-color: ${({ foundation, color = 'bg-white-normal' }) => foundation?.theme?.[color]};
    border-radius: 50%;
  }
`

export const LockIcon = styled(Icon)`
  ${absoluteCenter('')}
  z-index: 1;
`
