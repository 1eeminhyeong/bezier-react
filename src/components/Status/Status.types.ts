export enum StatusType {
  Online = 'Online',
  Offline = 'Offline',
  Lock = 'Lock',
}

export interface StatusProps {
  type: StatusType
}
