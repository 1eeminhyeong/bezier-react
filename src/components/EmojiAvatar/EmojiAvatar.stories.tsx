/* External dependencies */
import React from 'react'
import base from 'paths.macro'
import { Meta, Story } from '@storybook/react'

/* Internal dependencies */
import { getTitle } from 'Utils/storyUtils'
import { StatusType } from 'Components/Status'
import EmojiAvatar from './EmojiAvatar'
import EmojiAvatarProps, { EmojiAvatarSize } from './EmojiAvatar.types'

const MOCK_EMOJI_URL = 'https://cf.exp.channel.io/asset/emoji/images/80/blush.png'

export default {
  title: getTitle(base),
  component: EmojiAvatar,
  argTypes: {
    avatarUrl: {
      control: {
        type: 'text',
      },
    },
  },
} as Meta

const Template: Story<EmojiAvatarProps> = (args) => <EmojiAvatar {...args} />

export const Primary = Template.bind({})

Primary.args = {
  emojiUrl: MOCK_EMOJI_URL,
}

export const WithStatus = Template.bind({})

WithStatus.args = {
  emojiUrl: MOCK_EMOJI_URL,
  size: EmojiAvatarSize.Size30,
  status: StatusType.Online,
}
