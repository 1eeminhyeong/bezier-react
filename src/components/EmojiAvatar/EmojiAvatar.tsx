/* External dependencies */
import React, {
  useCallback,
  useMemo,
} from 'react'
import {
  noop,
  isEmpty,
  isNil,
} from 'lodash-es'

/* Internal dependencies */
import { Status } from 'Components/Status'
import EmojiAvatarProps, { EmojiAvatarSize } from './EmojiAvatar.types'
import {
  StatusWrapper,
  Wrapper,
  EmojiIcon,
} from './EmojiAvatar.styled'

export const EMOJI_AVATAR_TEST_ID = 'bezier-react-emoji-avatar'

function EmojiAvatar({
  emojiUrl,
  className,
  interpolation,
  status,
  size = EmojiAvatarSize.Size24,
  testId = EMOJI_AVATAR_TEST_ID,
  onClick = noop,
  children,
}: EmojiAvatarProps) {
  const handleClickIcon = useCallback((event: React.MouseEvent<HTMLDivElement>) => {
    if (!isNil(onClick)) {
      onClick(event)
    }
  }, [onClick])

  const StatusElement = useMemo(() => {
    if (
      (isEmpty(children) && !status)
      || (children && !React.isValidElement(children))
    ) {
      return null
    }

    const Contents = (() => {
      if (children) { return children }
      if (status) {
        return (
          <Status type={status} />
        )
      }
      return null
    })()

    return (
      <StatusWrapper>
        { Contents }
      </StatusWrapper>
    )
  }, [
    status,
    children,
  ])

  return (
    <Wrapper
      data-testid={testId}
      onClick={handleClickIcon}
    >
      <EmojiIcon
        role="img"
        imageUrl={emojiUrl}
        size={size}
        className={className}
        interpolation={interpolation}
        unClickable={isNil(onClick)}
      >
        { StatusElement }
      </EmojiIcon>
    </Wrapper>
  )
}

export default EmojiAvatar
