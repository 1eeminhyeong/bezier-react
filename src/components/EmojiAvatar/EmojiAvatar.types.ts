/* Internal dependencies */
import type {
  BezierComponentProps,
  ChildrenProps,
  SizeProps,
} from 'Types/ComponentProps'
import { StatusType } from 'Components/Status'

export enum EmojiAvatarSize {
  Size16 = 16,
  Size20 = 20,
  Size24 = 24,
  Size30 = 30,
  Size42 = 42,
  Size90 = 90,
  Size120 = 120,
}

interface EmojiAvatarOptions {
  emojiUrl: string
  status?: StatusType
  onClick?: React.MouseEventHandler<HTMLDivElement>
}

export default interface EmojiAvatarProps extends
  BezierComponentProps,
  ChildrenProps,
  SizeProps<EmojiAvatarSize>,
  EmojiAvatarOptions {}
