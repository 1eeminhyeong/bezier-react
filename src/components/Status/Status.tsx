/* External dependencies */
import React from 'react'

/* Internal denpendencies */
import { IconSize } from 'Components/Icon'
import { StatusType, StatusProps } from './Status.types'
import { StatusCircle, LockIcon } from './Status.styled'

// TODO: 테스트 코드 작성
const STATUS_TEST_ID = 'bezier-react-status'

function Status({
  type,
}: StatusProps) {
  if (type === StatusType.Lock) {
    return (
      <StatusCircle
        data-testid={STATUS_TEST_ID}
        color="bg-white-normal"
      >
        <LockIcon
          name="lock"
          size={IconSize.XXS}
          color="txt-black-darker"
        />
      </StatusCircle>
    )
  }

  return (
    <StatusCircle
      data-testid={STATUS_TEST_ID}
      color={type === StatusType.Online ? 'bgtxt-green-normal' : 'bg-black-dark'}
    />
  )
}

export default Status
