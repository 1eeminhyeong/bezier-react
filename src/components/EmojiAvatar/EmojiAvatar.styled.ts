/* Internal dependencies */
import {
  styled,
  css,
  absoluteCenter,
  centeredBackgroundImage,
} from 'Foundation'
import { InterpolationProps } from 'Types/Foundation'
import { EmojiAvatarSize } from './EmojiAvatar.types'

const AVATAR_STATUS_GAP = 2
const WRAPPER_SIZE = 24

interface IconProps extends InterpolationProps {
  unClickable?: boolean
  size?: EmojiAvatarSize
  privateChat?: boolean
}

export const Wrapper = styled.div`
  position: relative;
  top: calc(50% - ${WRAPPER_SIZE}px/2);
  left: calc(50% - ${WRAPPER_SIZE}px/2);
  box-sizing: content-box;
  display: flex;
  flex-shrink: 0;
  align-items: center;
  justify-content: center;
  width: ${WRAPPER_SIZE}px;
  height: ${WRAPPER_SIZE}px;
`

export const EmojiIcon = styled.div<IconProps>`
  width: ${({ size }) => size}px;
  height: ${({ size }) => size}px;
  cursor: pointer;
  border-radius: ${({ foundation }) => foundation?.rounding.round4};
  transition: ${({ foundation }) => foundation?.transition.getTransitionsCSS('background-color')};

  &:hover {
    background-color: ${({ foundation }) => foundation?.theme['bg-black-light']};
  }

  ${({ unClickable }) => unClickable && css`
    cursor: initial;
  `}

  ${absoluteCenter('')}

  ${({ size }) => css`
    ${centeredBackgroundImage(size, size)}
  `}

  ${({ interpolation }) => interpolation}
`

export const StatusWrapper = styled.div`
  position: absolute;
  right: -${AVATAR_STATUS_GAP}px;
  bottom: -${AVATAR_STATUS_GAP}px;
`
