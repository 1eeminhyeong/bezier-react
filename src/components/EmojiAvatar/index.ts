/* Internal dependencies */
import EmojiAvatar from './EmojiAvatar'
import EmojiAvatarProps, { EmojiAvatarSize } from './EmojiAvatar.types'

export type {
  EmojiAvatarProps,
}

export {
  EmojiAvatar,
  EmojiAvatarSize,
}
