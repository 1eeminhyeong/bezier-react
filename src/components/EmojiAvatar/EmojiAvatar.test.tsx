/* External dependencies */
import React from 'react'

/* Interanal dependencies */
import { render } from 'Utils/testUtils'
import { StatusType } from 'Components/Status'
import type EmojiAvatarProps from './EmojiAvatar.types'
import EmojiAvatar, { EMOJI_AVATAR_TEST_ID } from './EmojiAvatar'
import { EmojiAvatarSize } from './EmojiAvatar.types'

const MOCK_EMOJI_URL = 'https://cf.exp.channel.io/asset/emoji/images/80/a.png'

describe('EmojiAvatar Test >', () => {
  const defaultProps: EmojiAvatarProps = { emojiUrl: MOCK_EMOJI_URL }

  const renderComponent = (props?: Partial<EmojiAvatarProps>) => render(
    <EmojiAvatar {...defaultProps} {...props} />,
  )

  it('snapshot', () => {
    const { getByTestId } = renderComponent()
    const emojiAvatar = getByTestId(EMOJI_AVATAR_TEST_ID)

    expect(emojiAvatar).toMatchSnapshot()
  })

  it('should render default wrapper style', () => {
    const { getByTestId } = renderComponent()
    const emojiAvatar = getByTestId(EMOJI_AVATAR_TEST_ID)

    expect(emojiAvatar).toHaveStyle('position: relative')
    expect(emojiAvatar).toHaveStyle('width: 24px')
    expect(emojiAvatar).toHaveStyle('height: 24px')
    expect(emojiAvatar).toHaveStyle('left: calc(50% - 24px/2)')
    expect(emojiAvatar).toHaveStyle('top: calc(50% - 24px/2)')
    expect(emojiAvatar).toHaveStyle('border-radius: 4px')
  })

  it('should render Icon style', () => {
    const { getByTestId } = renderComponent()
    const emojiAvatar = getByTestId(EMOJI_AVATAR_TEST_ID)

    expect(emojiAvatar.children.length).toBe(1)

    expect(emojiAvatar.children.item(0)).toHaveStyle('width: 24px')
    expect(emojiAvatar.children.item(0)).toHaveStyle('height: 24px')
    expect(emojiAvatar.children.item(0)).toHaveStyle('position: absolute')
    expect(emojiAvatar.children.item(0)).toHaveStyle('top: 50%')
    expect(emojiAvatar.children.item(0)).toHaveStyle('left: 50%')
    expect(emojiAvatar.children.item(0)).toHaveStyle('transform: translate(-50%,-50%)')
    expect(emojiAvatar.children.item(0)).toHaveStyle(`background-image: url(${MOCK_EMOJI_URL})`)
  })

  it('should render status component with status prop', () => {
    const { getByTestId } = renderComponent({ status: StatusType.Lock })
    const emojiAvatar = getByTestId(EMOJI_AVATAR_TEST_ID)

    expect(emojiAvatar.children.item(0)?.children.item(0)?.children.item(0)).toHaveAttribute('data-testid', 'bezier-react-status')
    expect(emojiAvatar.children.item(0)?.children.item(0)).toHaveStyle('position: absolute')
    expect(emojiAvatar.children.item(0)?.children.item(0)).toHaveStyle('right: -2px')
    expect(emojiAvatar.children.item(0)?.children.item(0)).toHaveStyle('bottom: -2px')
  })

  it('should render children component without status prop', () => {
    const children = (
      <div id="test" style={{ width: '300px' }}>
        <div>happy</div>
      </div>
    )

    const { getByTestId } = renderComponent({ children })
    const emojiAvatar = getByTestId(EMOJI_AVATAR_TEST_ID)

    expect(emojiAvatar.children.item(0)?.children.item(0)?.children.item(0)?.id).toBe('test')
    expect(emojiAvatar.children.item(0)?.children.item(0)?.children.item(0)).toHaveStyle('width: 300px')
    expect(emojiAvatar.children.item(0)?.children.item(0)?.children.item(0)).toHaveTextContent('happy')
  })

  it('should render emoji icon with size prop - size20', () => {
    const { getByTestId } = renderComponent({ size: EmojiAvatarSize.Size16 })
    const emojiAvatar = getByTestId(EMOJI_AVATAR_TEST_ID)

    expect(emojiAvatar.children.item(0)).toHaveStyle('width: 16px')
    expect(emojiAvatar.children.item(0)).toHaveStyle('height: 16px')
  })

  it('should render emoji icon with size prop - size42', () => {
    const { getByTestId } = renderComponent({ size: EmojiAvatarSize.Size42 })
    const emojiAvatar = getByTestId(EMOJI_AVATAR_TEST_ID)

    expect(emojiAvatar.children.item(0)).toHaveStyle('width: 42px')
    expect(emojiAvatar.children.item(0)).toHaveStyle('height: 42px')
  })

  it('should render emoji icon with size prop - size120', () => {
    const { getByTestId } = renderComponent({ size: EmojiAvatarSize.Size120 })
    const emojiAvatar = getByTestId(EMOJI_AVATAR_TEST_ID)

    expect(emojiAvatar.children.item(0)).toHaveStyle('width: 120px')
    expect(emojiAvatar.children.item(0)).toHaveStyle('height: 120px')
  })
})
