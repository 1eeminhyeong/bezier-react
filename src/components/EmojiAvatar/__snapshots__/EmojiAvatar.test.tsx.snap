// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmojiAvatar Test > snapshot 1`] = `
.c0 {
  position: relative;
  top: calc(50% - 24px/2);
  left: calc(50% - 24px/2);
  box-sizing: content-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 24px;
  height: 24px;
}

.c1 {
  width: 24px;
  height: 24px;
  cursor: pointer;
  border-radius: overflow:hidden;
  border-radius: 4px;
  -webkit-transition: transition-delay:0ms;
  transition: transition-delay:0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: background-color;
  transition-property: background-color;
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  background-image: url(https://cf.exp.channel.io/asset/emoji/images/80/a.png);
  background-position: center;
  background-repeat: no-repeat;
  background-size: 24px,24px;
  background-clip: content-box;
}

.c1:hover {
  background-color: #00000014;
}

<div
  class="c0"
  data-testid="bezier-react-emoji-avatar"
>
  <div
    class="c1"
    role="img"
    size="24"
  />
</div>
`;
